implemented motif search

ymoon06 opened the pull request:
### Description

takes motif, read, min_overlap and optional parameter and performs  a motif search in the read. 
It returns a list of tuple (start of overlap in the read, fraction of overlap). uses list comprehension for fast execution. 
Used == for fast string comparison.
test files are also designed to test all possible corner cases.

Fixes # (issue)

### Type of change

Please delete options that are not relevant.
I am not sure what this part is about. But I can say I don't have bugs in the unit test file.(I fixed the bug along I code bits by bits)

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality
      to not work as expected)
- [ ] Documentation update

### Checklist

Please carefully read these items and tick them off if the statements are true
_or do not apply_.

- [x] I have performed a self-review of my own code
- [x] My code follows the existing coding style, lints and generates no new
      warnings
- [x] I have added type annotations to all function/method signatures, and I
      have added type annotations for any local variables that are non-trivial,
      potentially ambiguous or might otherwise benefit from explicit typing.
- [x] I have commented my code in hard-to-understand areas
- [ ] I have added ["Google-style docstrings"] to all new modules, classes,
      methods/functions or updated previously existing ones
- [x] I have added tests that prove my fix is effective or that my feature
      works
- [x] New and existing unit tests pass locally with my changes and I have not
      reduced the code coverage relative to the previous state
- [ ] I have updated any sections of the app's documentation that are affected
      by the proposed changes

If for some reason you are unable to tick off all boxes, please leave a
comment explaining the issue you are facing so that we can work on it
together.

ymoon06 committed:
implemented motif search

takes motif, read, min_overlap and optional paramter. uses list comprehension for fast excution. used == for fast comparison.
test files are also designed to test all possible corner cases.

ymoon06 committed:
changed position

changed position

ymoon06 committed:
changed location of file

changed location of file

ymoon06 committed:
changed location of file

ymoon06 committed:
changed comment format

ymoon06 committed:
follow flake8

follow flake8

ymoon06 committed:
changed format

and deleted test motif file

ymoon06 committed:
deleted

deleted

ymoon06 committed:
fixed everything related to flake8

ymoon06 commented:
fixed my style so that it passes flake8 tests. Seems like passed successfully. 
tested functions with all corner cases I can think of. 
1 more test to do is to test with a lot of reads and motifs so that we can examine how it performs under stressful condition.
Thank you.

mzavolan commented on read_motif_v6.py:
```diff
@@ -0,0 +1,42 @@
+
+@author: YB Moon (ymoon06)
+"""
+# assume motif <read
```
Provide a description of the input-output behavior of the function and of all parameters.

mzavolan commented on read_motif_v6.py:
```diff
@@ -0,0 +1,42 @@
+@author: YB Moon (ymoon06)
+"""
+# assume motif <read
+# use_N is for further implement as a parameter
```
What is use_N?

mzavolan commented on read_motif_v6.py:
```diff
@@ -0,0 +1,42 @@
+        not isinstance(read, str) or\
+        not isinstance(min_overlap, int) or\
+        not isinstance(use_n, bool):
+        raise TypeError('pass the right argument')
```
Message should be more precise, e.g. 'Incorrect argument type: [argument]'

mzavolan replied:
Don't you want to check that the strings are both in upper or lower case?

mzavolan commented on read_motif_v6.py:
```diff
@@ -0,0 +1,42 @@
+        raise TypeError('pass the right argument')
+    # passing min_overlap <1 does not make sense
+    if len(motif)==0 or min_overlap <1:
+        raise ValueError('pass the right value')
```
Make message more precise

mzavolan commented on read_motif_v6.py:
```diff
@@ -0,0 +1,42 @@
+    #list comprehension is faster. parallelizing is allowed
+    partial_overlaps_start=[]
+    if len(read)>= min_overlap:
+        #doesnt make sense to do len(motif)+1 because then it would be the 2nd case
```
Please try to keep the comments concise. They should be readable to someone that may want to pick up the code and develop it further.

mzavolan commented on read_motif_v6.py:
```diff
@@ -0,0 +1,42 @@
+    # compute matches(overlaps) of the motif inside the read (2nd case)
+    # you compare the full motif length
+    full_overlaps=[]
+    if len(read)>-len(motif):
```
Is this test necessary? lengths will be >= 0

mzavolan commented on read_motif_v6.py:
```diff
@@ -0,0 +1,42 @@
+"""
+# assume motif <read
+# use_N is for further implement as a parameter
+def find_overlaps(motif, read, min_overlap, use_n=False):
```
Can you please add an option to report only matches that are fully contained in the read?

mzavolan commented on test_motif_shortened.py:
```diff
@@ -0,0 +1,83 @@
+
+if __name__ == '__main__':
+  unittest.main()
+  #print('success')
```
add new line at the end of the line

mzavolan commented on test_motif_shortened.py:
```diff
@@ -0,0 +1,83 @@
+    self.assertTrue(len(rv) == 2)
+    self.assertTrue(rv[0] == (3, 1))
+    self.assertTrue(rv[1] == (9, 2/5))
+
```
Can you implement the situation of no match whatsoever?

ymoon06 committed:
followed some feedbacks

followed some feedbacks

ymoon06 commented:
I didn't upload test file because it had lots of errors with respect to flake8 style.
Don't know how they are visible because I didnt upload it it.
But I think I fixed everything you requested with respect to the motif find function.
If time allowed, I will fix test file as well and upload it separately.
Thank you.

uniqueg committed:
Merge branch 'dev' into dev

uniqueg commented on read_motif_v6.py:
```diff
@@ -0,0 +1,73 @@
+        raise ValueError('motif should be all big characters')
+    if not read.isupper():
+        raise ValueError('read should be all big characters')
+    """compute partial overlaps of the motif at the start of the read"""
```
Use `"""` comments only for docstrings, otherwise use `#` comments. Docstrings are the documentation strings that are expected as the first statement of a module, class, function and method.

uniqueg commented on read_motif_v6.py:
```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+"""
```
Please only add a _single_ docstring. Also, please omit the author information, this is automatically tracked by Git, on a line-per-line basis. Adding this here will lead to artifacts, because these metadata are never properly updated when maintaining the code

uniqueg replied:
Please have a look at and follow the style of Google-style docstrings: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

uniqueg commented on read_motif_v6.py:
```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
```
Please add your code to the `htsinfer/infer_read_layout.py` module rather than its own module

uniqueg commented on read_motif_v6.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def find_overlaps(motif, read, min_overlap, full_contain=False):
+    """check the type of arguments """
```
Please add a complete Google-style docstrings: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

uniqueg commented on read_motif_v6.py:
```diff
@@ -0,0 +1,73 @@
+"""
+
+
+def find_overlaps(motif, read, min_overlap, full_contain=False):
```
Please add type hints to indicate the data types of the input parameters as well as the return type: https://docs.python.org/3.8/library/typing.html

ymoon06 committed:
docstring

docstring, moved position

ymoon06 committed:
Merge branch 'dev' of https://github.com/ymoon06/htsinfer into dev

ymoon06 committed:
reduced if

reduced if

ymoon06 committed:
fixed

fixed

ymoon06 committed:
changedq

changed

ymoon06 committed:
fix

fix

ymoon06 committed:
changed position

putting it into hstinfer directory causes coverage decrease without reason

ymoon06 commented:
fixed everything except one. If I put my file into the directory then the coverage decreases to 84% or so without any reason. if I put it outside then coverage does not decrease. do you have any idea on this? It is the same code and it is because of the error of my code.

ymoon06 committed:
upload test file

upload test file and changed directory

ymoon06 committed:
final change

final change

ymoon06 committed:
change

change

ymoon06 committed:
changed

changed

ymoon06 committed:
final change

final change

ymoon06 commented:
fixed everything, added a separate test file testing all that I have implemented including arguments.
One thing is that I didnt include it inside the existing file because it caused an build error.
May I ask if it is ok?

uniqueg commented:
It will be fine like that, thank you very much! If @zavolan is still happy with the logic (I think she was), I will make any final changes, if necessary, and merge. Might take a few days though

ymoon06 committed:
Merge branch 'dev' into dev

uniqueg committed:
Merge branch 'dev' into dev

mzavolan reviewed (approved):
Type checks could be done via mypy/linter.
